### ISSUE TITLE
Fix weightedExpertSum across singleton and multi-expert compiled calls

### ISSUE THREAD

**Issue by ronaldmannak** [PicoMLX/mlx-swift-lm#51]
A standalone GPU reproduction in the currently pinned integration revision `36983e7df71d9d6709b8ed4d9d387f23bc94446e` shows incorrect results from `MLXLMCommon.weightedExpertSum` after the compiled function first sees a singleton expert axis and then a larger expert axis in the same process.

This was isolated while validating Qwen MTP in PicoCore. It reproduces without a model, MTP, caches or PicoCore:

```swift
import MLX
import MLXLMCommon

Device.withDefaultDevice(.gpu) {
    for k in [1, 4, 2, 4] {
        let values = MLXArray(Array(1...(17 * k * 16)))
            .asType(.float32).reshaped([1, 17, k, 16])
        let weights = MLXArray.ones([1, 17, k]) / Float(k)
        let actual = weightedExpertSum(values, weights)
        let expected = (values * weights.expandedDimensions(axis: -1)).sum(axis: -2)
        eval(actual, expected)
        print(k, actual.shape, abs(actual - expected).max().item(Float.self))
    }
}
```

Observed in a fresh test process on Apple silicon:

- k=1: shape [1,17,16], maximum absolute difference 0.
- k=4: same shape, difference 996.
- k=2: same shape, difference 400.
- k=4 again: difference 996.

In the model-level fixture this also produced inconsistent full-prefill logits and incorrect greedy output when a one-expert Qwen MoE fixture preceded a four-expert fixture. A fresh process running only the four-expert fixture agreed with ordinary chunked generation. Direct `SwitchGLU` full-versus-per-row output agreed within 4.9e-8, so its sorting permutation is not implicated by this reproduction. Training mode and mixed CPU/GPU execution were investigated and do not explain the standalone result above.

`weightedExpertSum` in `Libraries/MLXLMCommon/SwitchLayers.swift` uses `compile(shapeless: true)`. A singleton reduction can be simplified during tracing and reused for a non-singleton dimension. That is the suspected mechanism; the numerical failure above is directly observed.

Suggested minimal fix to validate: make this reduction shape-specialized (`shapeless: false`) or keep singleton expert reductions outside the shared compiled function. Add this regression to mlx-swift-lm, including 1→4 and 4→1 order, repeated calls, and the normal multi-expert case. Also inspect other callers of this shared function before deciding whether a broader MLX compiler issue needs a separate report. Do not add a PicoCore copy or workaround of the native reduction.

PicoCore can proceed with dense Qwen MTP. Qwen MoE MTP remains explicitly rejected pending the native fix and parity validation. No dependency code or pins were changed during this investigation. This issue is in our fork for focused follow-up; no comment or issue was posted upstream.

**Comment by ronaldmannak** [comment]
Implemented in 8e3b33261bcb86a2c7bc236287ec6bfc631d725f on the existing integration branch, tracked by PR #32 under PicoCore #152. The shape-specialized native reduction passes the singleton/multiple expert regression in both call orders across float32/float16/bfloat16. The regression fails on the previous implementation; the full native suite passes. PicoCore MoE integration is PR https://github.com/PicoMLX/PicoCore/pull/166, which now has current-head Codex thumbs-up and green CI. No PicoCore reduction copy was added.

Closing the fork prerequisite because the tested implementation is published through the existing integration PR; fork main remains the upstream mirror. Current integration head f108b020 has Codex thumbs-up and green macOS CI. The final runtime tree passed 802 Swift Testing tests and 618 XCTest cases, with two existing skips.

AI assistance: implementation, validation and this status update prepared with Codex under the maintainer’s authorization.
